lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +14,241 @@
+// === Constants ===
+// =================
+
+mod constants {
+    /// The number of 'frames' to reserve in the state stack.
+    pub const STATE_STACK_RESERVATION:usize = 1024;
+    /// The size of the output buffer (in tokens) to reserve.
+    pub const OUTPUT_BUFFER_RESERVATION:usize = 1024;
+}
+
+
+
+// ==============
+// === Flexer ===
+// ==============
 
 /// The flexer is an engine for generating lexers.
 ///
 /// Akin to flex and other lexer generators, it is given a definition as a series of rules from
 /// which it then generates code for a highly optimised lexer implemented on top of a
 /// [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
-pub trait Flexer {
+///
+/// Lexers defined using the flexer work on a stack of _states_, where a state is represented by a
+/// [`crate::group::Group`]. Being in a given state (represented below by the top of the
+/// `state_stack`) means that the flexer can match a certain set of rules associated with that
+/// state. The user may cause the lexer to transition between states by pushing and popping states
+/// on the stack, thus allowing a much more flexible lexing engine than pure regular grammars.
+#[derive(Clone,Debug)]
+pub struct Flexer<Definition,Output,Reader> {
+    /// The stack of states that are active during lexer execution.
+    pub state_stack: Vec<group::Identifier>,
+    /// A reader for the input.
+    pub reader: Reader,
+    /// The result of the current stage of the DFA.
+    pub status: StageStatus,
+    /// The tokens that have been lexed.
+    pub output: Vec<Output>,
+    /// The text of the current match of the lexer.
+    pub current_match: String,
+    /// The definition of the user-provided state for the lexer.
+    definition: Definition,
+}
+
+impl<Definition,Output,Reader> Flexer<Definition,Output,Reader>
+where Definition:State, Reader:LazyReader {
+    /// Creates a new lexer instance.
+    pub fn new(mut reader:Reader) -> Flexer<Definition,Output,Reader> {
+        let mut state_stack = Vec::default();
+        state_stack.reserve(constants::STATE_STACK_RESERVATION);
+        let status     = default();
+        let mut output = Vec::default();
+        output.reserve(constants::OUTPUT_BUFFER_RESERVATION);
+        let definition       = Definition::new(&mut reader);
+        let initial_state_id = definition.initial_state();
+        state_stack.push(initial_state_id);
+        let current_match = default();
+
+        Flexer{state_stack,reader,status,output,definition,current_match}
+    }
+}
+
+/// This block is things that are part of the lexer's interface and functionality.
+impl<Definition,Reader,Output> Flexer<Definition,Output,Reader>
+where Definition: State, Output:Clone {
+    /// Gets the lexer result.
+    pub fn get_result(&mut self) -> &Vec<Output> {
+        &self.output
+    }
+
+    /// Gets the lexer's root state.
+    pub fn root_state(&self) -> group::Identifier {
+        self.definition.initial_state()
```

Comment:
I don't like that we are using two names for the same thing - `root_state` and `initial_state`. Why don't we just keep one of them?